server/ladder_service/ladder_service.py

```diff
@@ -519,6 +563,66 @@ def make_game_options(player: Player) -> GameLaunchOptions:
                     player.state = PlayerState.IDLE
                 player.write_message(msg)
 
+            if abandoning_players:
+                self._logger.info(
+                    "Players failed to connect: %s",
+                    abandoning_players
+                )
+                self.violation_service.register_violations(abandoning_players)
+
+    async def launch_match(
+        self,
+        game: LadderGame,
+        host: Player,
+        guests: list[Player],
+        make_game_options: Callable[[Player], GameLaunchOptions]
+    ):
+        # Launch the host
+        if host.lobby_connection is None:
+            raise NotConnectedError([host])
+
+        host.lobby_connection.write_launch_game(
+            game,
+            is_host=True,
+            options=make_game_options(host)
+        )
+
+        try:
+            await game.wait_hosted(60)
+        except asyncio.TimeoutError:
+            raise NotConnectedError([host])
+        finally:
+            # TODO: Once the client supports `match_cancelled`, don't
+            # send `launch_game` to the client if the host timed out. Until
+            # then, failing to send `launch_game` will cause the client to
+            # think it is searching for ladder, even though the server has
+            # already removed it from the queue.
+
+            # Launch the guests
+            not_connected_guests = [
+                player for player in guests
+                if player.lobby_connection is None
+            ]
+            if not_connected_guests:
+                raise NotConnectedError(not_connected_guests)
+
+            for guest in guests:
+                assert guest.lobby_connection is not None
+
+                guest.lobby_connection.write_launch_game(
+                    game,
+                    is_host=False,
+                    options=make_game_options(guest)
+                )
+        try:
+            await game.wait_launched(60 + 10 * len(guests))
+        except asyncio.TimeoutError:
+            connected_players = game.get_connected_players()
```

Comment:
I don't understand when we set a player as connected. It seems that we need to call `set_player_option()` at some point to mark him as connected, but I can't find when this would actually happen.

Reply:
It's done when the game host sends the `PlayerOption` GPGNet command (through GameConnection). This is how we know that the host has successfully established a connection to that player.